### DISCUSSION TITLE
Help with Door sensors? Can anyone explain a few basics to me.

### DISCUSSION THREAD

**Discussion by dali-dbr** [arendst/Tasmota#21532] [Q&A] [answered]
Can anyone please explain a few basic things to me?  I'm confused by reading all sorts of info about Esp32 on the internet. I would really like to clear some things up before I print the PCB board.

![image](https://github.com/arendst/Tasmota/assets/6317229/b1bce551-e106-4ca4-9452-bed751563095)

I've installed Tasmota 1.4 on 38pin ESP32-DevKit board.  I used 14 available GPIO's for magnetic sensors. GPIOs --> GND
**All inputs work great, simultaneously.** I also tested it with over 100 meters long Cat6 cable and still works nicely.

![image](https://github.com/arendst/Tasmota/assets/6317229/cca97352-1767-4103-a4a6-b2f49c2a4697)
![image](https://github.com/arendst/Tasmota/assets/6317229/96d062e8-f4aa-4063-b1f7-9f059392eb09)

Please help me clarify some things.

**1. GPIOs safe to use or to be avoided**
I read that there are some GPIOS that should be avoided, else the ESP32 might not boot if in the wrong state.  Howerver, I tried with all 14 GPIOs Low or High and it boots every time.

**2. Pull up / Pull down resistors?**
I read a lot about pull up/pull down resistors? On my breadboard all 14 gpios (using the magnetic door sensors) are working nicely without any resistors.  I also read that some GPIOs already have pull up/pull down resistors so I'm really lost here. Where or how to check this for specific ESP32 board ? For **wt32-eth01** board for example?

**Other wiring for the magnetic door sensors**
I can also find other wirings. Sending 3,3V to to the sensor and back to GPIO's.
![image](https://github.com/arendst/Tasmota/assets/6317229/75ee5d18-ef93-4cc8-85a5-8a6acb3683ac)

I'm a bit confused by all this, but I would like to get it done properly, hopefully to avoid any issues later on.  I would be very grateful If someone can help me with my confusion over things I do not yet fully understand. 

Thank you

**Comment by sfromis** [comment]
Standard behavior when configuring as a Switch (or Button) gpio is that a weak internal pullup is activated, working well when you have the door sensor connect the gpio to ground. With long wiring, there can be a risk of that pullup not being strong enough, in case of trouble (including noise) you *may* have to add a stronger external pullup, like 4.7KΩ

I'd recommend against connecting the door switches to 3.3V, as this is not the standard way of wiring such stuff. Not impossible if you have specific reasons, but having voltage on the lines does increase risks of unfortunate events, or attenuation with long wires.

If you use gpio 0 for a door sensor, and it is activated (closed circuit) when you boot, that will cause trouble by having the chip start in bootloader mode instead of booting Tasmota. The pin is already connected to the boot button. Not recommended.

Gpio 2 is also a strapping pin, but only when you want the bootloader mode, I'd still avoid that pin if having others available, also because it is often connected to a led on the board, or having an external pullup already.

Gpio 12/15 can also have an effect at boot time, meaning that I again prefer to just use other pins. Not saying that they *will* cause trouble, I just prefer unencumbered pins.

While gpio 1/3 are not strapping pins, they are connected to the onboard USB/serial adapter chip, and better avoided for other purposes.

**Reply by dali-dbr** [reply]
Thank you VERY much for this valuable info!!!

I'm having trouble understanding where exactly in my case this 4.7KΩ external pullup should be placed. Can you please update my drawing if not too much to ask?
![image](https://github.com/arendst/Tasmota/assets/6317229/e90e8632-21e9-48f2-8073-19d74c6a11a1)

Based on what detail you can tell which GPO's are suitable and which is better to avoid? I would like to understand myself to be able to check for other ESP boards as well....

https://www.studiopieters.nl/esp32-pinout/
![image](https://github.com/arendst/Tasmota/assets/6317229/c6ef6d8b-fcea-40f4-9745-83a4b07b2ae7)

Ok, so I exclude 12 and 15 as well and I get 12 safe GPIO's
![image](https://github.com/arendst/Tasmota/assets/6317229/4aa0db24-248c-4c52-9749-48c18cd9fb9a)

Please note I use this Esp32 board just for testing purposes. Later I intend to use **wt32-eth01** cause of ethernet LAN port,(sadly not POE). I believe it is compatible with Tasmota 1.4. ¸

I was thinking to use 10 GPIOs. Is there any GPO better to avoid?
![image](https://github.com/arendst/Tasmota/assets/6317229/a8ba4721-a0b7-4224-8aa1-d2ca6e49a435)

Thank you very much for your help!!!

**Reply by sfromis** [reply]
If you need a pullup, which is not a requirement, it would be between the signal (in this case G19) and 3V3.

I'd suggest reading this page about gpio pin selection: https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

WT32-ETH01 is a much more complex board to deal with. Just because some pin is broken out does not mean that it is available, due to the Ethernet feature using many fixed gpios, which are not visible in the template configuration. I've no experience working with the pins there, but the comment block here does mention "some numbers"....
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_82_esp32_ethernet.ino#L23-L65
(Of course, without an onboard USB/serial adapter chip, that does relax the previously mentioned considerations a bit).

I can't really say that it is a board I'd select as a basis to build on, when many pins are needed. While I do have one, I've avoided using it for more than Ethernet testing.....

**Comment by dali-dbr** [comment]
Thank you.

So you recommend an input voltage of 3V3 instead of 5V?  I currently have designed as 5V AC/DC but I can change that.  

I ordered wt32-eth01, just for the safe of ethernet port, so I'll will play with it once get it. I will need to find out how to set static IP for Ethernet port, which I guess is not the same as for Wifi.

![image](https://github.com/arendst/Tasmota/assets/6317229/8d408ec9-c385-4da4-8224-07e16a192d48)


Can you recommend any more suitable ESP32 board with an ethernet port for about 10 door sensors?

Thank you

**Reply by sfromis** [reply]
There was no hint of such a recommendation in that I posted about the 3V3 pin. When you use a dev board, it will have a voltage regulator for input 5V, and producing 3.3V. My point was **only** that *if* you need a pullup, that needs to be from 3.3V, which is generated locally on the board. However, if you have a good stable 3.3V power source close to the board (not with long connecting wires), this is ok too, but you do skip the stabilization the voltage regulator circuit gives. Not sure about implications for Ethernet, but this for me is an argument for 5V.

When it comes to Ethernet commands.... See https://tasmota.github.io/docs/Commands/#ethernet
I generally prefer to **not** be using static IP addresses, but see "best practice" as being via DHCP. Obviously, circumstances means that YMMV.

When it comes to Ethernet support, newer ESP32 variants no longer has this as part of the ESP32 chip, and support for the W5500 Ethernet chip via a SPI bus has been added to Tasmota. This is simpler with only a few pins needed. However, my experience with Ethernet-supporting boards is too low for me to have recommendations on hand.

**Comment by dali-dbr** [comment]
Thank you.

To summarize it for anyone who might come across this post.

**3.3V or 5V?**
It's best to use 3.3V (Esp native input voltage) unless any other device in the circuit board requires 5V. 5V pin input uses a regulator to get it down to 3,3V and will only cause additional power consumption and heating for no good reason.

**Ethernet vs Wifi.?*
From my experience (from other fields), the cheapest cable is much more reliable than the most expensive wireless you can buy. For example, Shelly Professional series all have Ethernet. So if both options are available, I will always suggest using the cable.

**External stronger Pullup resistors?**
Some GPIOs have internal pullup resistors, so those do not require it. But having external stronger pullup resistors can not hurt either, therefore best to use those in any case.

Please correct me if I'm wrong

Thank you

**Reply by sfromis** [reply]
No, it is not "best" to use 3.3V to power the board, for the reasons I already mentioned. Just another option, including downsides.

And sure, if you already have easy Ethernet wiring on location, that's a good option, but only a tiny part of device placement have the convenience of such cabling easily available. It you have that, maybe consider a board with PoE too, to still only need one cable. And the Wifi needs of Tasmota is **very** low, meaning that nuances of stable reliability does not mean much.

And **all** gpios have internal pullup resistors, which you can choose to enable or disable. Those are weak, and in some circumstances you might need a stronger one. But no, it is not "best" to always add external pullups, in most cases is is a waste of space and work. And uses more power.....

The world is always more nuanced than simple one size fits all answers.

**Comment by dali-dbr** [comment]
Hi.

By power supply 3.3V I meant HLK-PM03 Hilink on the same PCB board. But I understand your point about the regulator. When DC source is away, there is always a power loss on cables, so it makes a perfect point to use 5V input and let the regulator do its job.

Regarding pullups. I'm not sure all GPIOs have pullups on Wemos D1 mini. Two GPIOS (I believe 15 and 16) didn't work without an external pullup. By the way.  Where in Tasmota can I choose to enable or disable pullups?

Also for Wemos D1 mini flashed with Tasmota 14 I don't get GPIO viewer. The menu Tools is not there. Is that only for Esp32 Devkit?

Thank you

**Reply by dali-dbr** [reply]
I agree with you on POE option. 

https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware

Do you think Tasmota will work with this board?

Thank you

**Reply by Jason2866** [reply]
Yes.

**Reply by sfromis** [reply, marked as answer]
When configuring input pins, you can configure Switch, or Switch_n for no pullup. Etc. On ESP8266, the GPIO Viewer is an optional feature you can enable if you do a custom build.
`#define USE_GPIO_VIEWER`

**Reply by dali-dbr** [reply]
Perfect.

**Reply by dali-dbr** [reply]
Hi.

So I got wt32-eth01 board and installed Tasmota 14.1.

I managed to get 8 GPIOS working with a magnetic DW sensor. I skipped IO1 as it is used for programming the device.

![image](https://github.com/arendst/Tasmota/assets/6317229/989df9cf-f698-4d92-ac52-47e5937cc58a)

From here on I have 2 questions.

**1. Pull up voltage and correct PIN?**
Maybe a stupid question but I'd rather ask anyway. What would be the correct PIN to connect pull up resistor to? All 3 works. **I'm assuming the RED 3v3 is correct.**  The green 3V3 will produce 3,4V and it works too. 5V I guess is too much for GPIO inputs.

![image](https://github.com/arendst/Tasmota/assets/6317229/54c590c5-25a9-4417-a98c-a4cc4ab36cc2)

**2. Pull up also on GPIOS that does not need it?**
GPIOS 35,36,39 needs a Pull-Up resistor. I used 4k7 as was suggested here. GPIOS 2,4,12,14,15 works without Pull-ups here on a breadboard.  In real use, I will use longer cables to reach the DW sensors. Since I will use pull-up for 3 GPIOS, I might include it for the other 5, unless it can cause some problems. I also read that some GPIOS can have weak internal pull-ups. So, does it hurt/doesn't matter/can improve having external Pull-ups?

Thank you very much!

Br

**Reply by Henkonoly** [reply]
Hi,

just wondering, is everithing working on a stable way now?
